controllers/cluster_controller.go

```diff
@@ -18,44 +18,98 @@ package controllers
-// the Cluster object against the actual cluster state, and then
-// perform operations to make the cluster state reflect the state specified by
-// the user.
-//
-// For more details, check Reconcile and its Result here:
-// - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.8.3/pkg/reconcile
 func (r *ClusterReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
-	_ = log.FromContext(ctx)
+	log := r.Log.WithValues("cluster", req.NamespacedName)
+
+	// Fetch the Cluster instance.
+	cluster := &capiv1alpha3.Cluster{}
+	if err := r.Client.Get(ctx, req.NamespacedName, cluster); err != nil {
+		if apierrors.IsNotFound(err) {
+			return ctrl.Result{}, nil
+		}
 
-	// your logic here
+		return ctrl.Result{}, err
+	}
 
-	return ctrl.Result{}, nil
+	return r.ReconcileDeletion(ctx, cluster, log)
+}
+
+func (r *ClusterReconciler) ReconcileDeletion(ctx context.Context, cluster *capiv1alpha3.Cluster, log logr.Logger) (ctrl.Result, error) {
+	if _, ok := cluster.Annotations[ignoreClusterDeletion]; ok {
+		log.Info(fmt.Sprintf("Found annotation %s. Cluster %s/%s will be ignored for deletion", ignoreClusterDeletion, cluster.Namespace, cluster.Name))
+		return ctrl.Result{}, nil
+
+	}
+
+	if !cluster.DeletionTimestamp.IsZero() {
+		log.Info(fmt.Sprintf("Deletion for cluster %s/%s is already applied", cluster.Namespace, cluster.Name))
+		return ctrl.Result{}, nil
+	}
+
+	// immediately delete the cluster if defaultTTL has passed
+	if deletionTimeReached(cluster) {
+		err := r.Client.Delete(ctx, cluster, client.PropagationPolicy(metav1.DeletePropagationBackground))
+		if err != nil {
+			log.Error(err, fmt.Sprintf("unable to delete cluster %s/%s", cluster.Namespace, cluster.Name))
+		}
+		log.Info(fmt.Sprintf("Cluster %s/%s will be deleted", cluster.Namespace, cluster.Name))
+
+		return ctrl.Result{}, nil
+	}
+
+	// only send marked for deletion event if we still have ~1h before the cluster gets deleted
+	if deletionEventTimeReached(cluster) {
+		log.Info(fmt.Sprintf("Cluster %s/%s is marked for deletion", cluster.Namespace, cluster.Name))
+		r.clusterDeletionEvent(cluster, fmt.Sprintf("Cluster %s/%s is marked for deletion", cluster.Namespace, cluster.Name))
+		return ctrl.Result{
+			RequeueAfter: 1 * time.Hour,
+		}, nil
+	}
+
+	return requeue(), nil
 }
 
 // SetupWithManager sets up the controller with the Manager.
 func (r *ClusterReconciler) SetupWithManager(mgr ctrl.Manager) error {
-	return ctrl.NewControllerManagedBy(mgr).
-		// Uncomment the following line adding a pointer to an instance of the controlled resource as an argument
-		// For().
+	err := ctrl.NewControllerManagedBy(mgr).
+		For(&capiv1alpha3.Cluster{}).
 		Complete(r)
+	if err != nil {
+		return errors.Wrap(err, "failed setting up with a controller manager")
+	}
+
+	r.recorder = mgr.GetEventRecorderFor("cluster-controller")
+	return nil
+}
+
+func (r *ClusterReconciler) clusterDeletionEvent(cluster *capiv1alpha3.Cluster, message string) {
```

Comment:
What do you think about renaming to `sendClusterDeletionEvent` or `submitClusterDeletionEvent` ?

Reply:
Good idea 👍🏻